### DISCUSSION TITLE
Reusing timestamps for TimeSeries objects in MatNWB?

### DISCUSSION THREAD

**Discussion by jccava** [NeurodataWithoutBorders/helpdesk#43] [Q&A]
Hi.  I am using MatNWB.   I have several channels of analog data that are all acquired using the same time values.  I can easily create multiple **TimeSeries** objects by setting **'data'** to the values acquired and **'timestamps'** to the times.  However, I noticed that PyNWB has the ability to reuse **timestamps** among **TimeSeries** objects:

```
data = list(range(101, 201, 10))
reuse_ts = TimeSeries(
    name="reusing_timeseries",
    data=data,
    unit="SIunit",
    timestamps=time_series_with_timestamps,
)

```
This tactic does not seem to work with MatNWB for me, and the **validate_timestamps** method for the **TimeSeries** object only appears to check for **float64**.  Am I missing something, or is this 'reusing' functionality not available in MatNWB?

**Comment by lawrence-mbf** [comment]
> this 'reusing' functionality not available in MatNWB

Yes, more advanced features like reusing is not present in MatNWB. I'm not entirely sure what goes on in "reuse" but you will have to just do that manually. I'm assuming there's more to reuse than just reassignment @rly?

**Reply by oruebel** [reply]
> I'm not entirely sure what goes on in "reuse"

Reuse in this case (if I remember correctly) means that PyNWB will create a link to the existing timestamps dataset instead of copying the timestamps. In most cases that would be an internal link within an NWB file, but I believe it could also be an external link. In MatNWB, I believe this means you might be able to do something along the lines of ``'timestamps', types.untyped.SoftLink(...)``. From a UI perspective, PyNWB accepts another TimeSeries as the timestamps in the constructor so that it can figure out where it will need to link to https://pynwb.readthedocs.io/en/stable/tutorials/general/file.html#reusing-timestamps

**Reply by jccava** [reply]
Yes!  I explored using SoftLink, but one cannot softlink to float64.  The set of SoftLink-able objects is not all-inclusive.  However, if VectorData were SoftLink-able, and the 'timestamps' property accepted VectorData....???

**Reply by jccava** [reply]
I am importing data for a conversion pipeline, so my first pass is use-agnostic, so I cannot put these data into a SpatialSeries until I decode their source.  But I'd love to save time and storage space.  If I just need to duplicate the 10 Million timestamps for each in Matlab, so be it.

**Reply by jccava** [reply]
I absolutely appreciate the thoughtful and informed replies here.  Thank you for that, even if it's not what I would like in a perfect NWB world. ;)

**Reply by lawrence-mbf** [reply]
To reference the `timestamps` dataset, you'd have to use `SoftLink` with a custom path. By magical convention, the HDF5 path should look something like `/acquisition/TimeSeries/timestamps` so just providing that object name should be enough.

Perhaps we could introduce a hybrid `SoftLink` constructor that takes an object and a property name so it can be bound this way.

**Reply by oruebel** [reply]
> Perhaps we could introduce a hybrid `SoftLink` constructor that takes an object and a property name so it can be bound this way.

This is effectively how PyNWB gets around the issue to be able to create links to data that has not been written to file yet. PyNWB accepts that full TimeSeries object as input for timestamps in that case assumes that we want to reuse the timestamps from that TimeSeries.

**Reply by jccava** [reply]
Fantastic information.  I'll give the custom softlink path a try and let you all know how I fare

**Reply by jccava** [reply]
Hey all!  I gave it a try.  It appears that the softlink works, but then the encoding tries to put more stuff under _timestamps/_, but _timestamps_ is apparently  already established by the softlink, so the code balks at adding existing data under _timestamps_ ?

I will send error messages if desired, but my traceback through the code points to this?

Thanks for all your help.  I'm including a unique vectordata for the times in each timeseries at the moment, so this is not work-stoppage.  But we do record a LOT of analog data on the same time units.

Please racall, this is MatNWB.  

Best,

James

**Reply by jccava** [reply]
To be clear, when I add additional analog data, and try to link to the softlinked times, that's when the code complains about things under _timestamps/_ already existing.

Thanks again.  james

**Reply by lawrence-mbf** [reply]
Please post the specific error message.

**Reply by jccava** [reply]
Working from home today.  I'll post the matlab complaint in the morning.

Sheesh.  You people are fast at responding!

James

**Reply by jccava** [reply]
Here's what happens:

```
Processing aiChannels
Processing convertInfo
Processing ops
Data from ops.mat not required
Processing rez
Data from rez.mat not required
Processing rez2
Data from rez2.mat not required
Processing sampsToSecsMap
Processing strobedEvents
Error using hdf5lib2
The HDF5 library encountered an error and produced the following stack trace information:

    H5G__traverse_real    component not found

Error in H5O.open (line 27)
output = H5ML.hdf5lib2('H5Oopen',loc_id,relname,lapl_id);

Error in io.writeAttribute (line 8)
oid = H5O.open(fid, path, 'H5P_DEFAULT');

Error in types.core.TimeSeries/export (line 407)
            io.writeAttribute(fid, [fullpath '/timestamps/interval'], obj.timestamps_interval);

Error in types.untyped.Set/export (line 181)
                    refs = v.export(fid, propfp, refs);

Error in types.core.NWBFile/export (line 843)
        refs = obj.acquisition.export(fid, [fullpath '/acquisition'], refs);

Error in NwbFile/export (line 62)
                refs = export@types.core.NWBFile(obj, output_file_id, '/', {});

Error in nwbExport (line 36)
    nwb(i).export(filename);

Error in processKilosortForNWB (line 92)
nwbExport(nwbOutputFile, nwbName);
 
```

**Reply by lawrence-mbf** [reply]
Yeah, it's trying to write with the assumption that timeseries exists as a dataset. That is a bug.

**Reply by jccava** [reply]
Do I get a lollipop???  Seriously, the work-around isn't all that bad.  Not work stoppage here.  But thanks for your attention to this.
